WooCommerce/Classes/ViewModels/CardPresentPayments/CardPresentModalServerSidePaymentCaptureError.swift

```diff
@@ -0,0 +1,83 @@
+import UIKit
+
+/// Modal presented on error capturing payment on the server side.
+final class CardPresentModalServerSidePaymentCaptureError: CardPresentPaymentsModalViewModel {
+    /// A closure to execute when the primary button is tapped.
+    private let primaryAction: () -> Void
+
+    /// A closure to execute after the secondary button is tapped to dismiss the modal.
+    private let dismissCompletion: () -> Void
+
+    let textMode: PaymentsModalTextMode = .reducedBottomInfo
+    let actionsMode: PaymentsModalActionsMode = .twoAction
+
+    let topTitle: String
+
+    var topSubtitle: String? = nil
+
+    let image: UIImage = .paymentErrorImage
+
+    let primaryButtonTitle: String?
+
+    let secondaryButtonTitle: String?
+
+    let auxiliaryButtonTitle: String? = nil
+
+    let bottomTitle: String?
+
+    let bottomSubtitle: String? = nil
+
+    var accessibilityLabel: String? {
+        guard let bottomTitle = bottomTitle else {
+            return topTitle
+        }
+        return topTitle + bottomTitle
+    }
+
+    init(primaryAction: @escaping () -> Void,
+         dismissCompletion: @escaping () -> Void) {
+        self.topTitle = Localization.title
+        self.bottomTitle = Localization.description
+        self.primaryButtonTitle = Localization.tryAgain
+        self.secondaryButtonTitle = Localization.noThanks
+        self.primaryAction = primaryAction
+        self.dismissCompletion = dismissCompletion
+    }
+
+    func didTapPrimaryButton(in viewController: UIViewController?) {
+        primaryAction()
+    }
+
+    func didTapSecondaryButton(in viewController: UIViewController?) {
+        viewController?.dismiss(animated: true) { [weak self] in
+            self?.dismissCompletion()
+        }
+    }
+
+    func didTapAuxiliaryButton(in viewController: UIViewController?) { }
+}
+
+private extension CardPresentModalServerSidePaymentCaptureError {
+    enum Localization {
+        static let title = NSLocalizedString(
+            "Payment submission failed",
+            comment: "Error title. Presented to users after collecting a payment in the app succeeds but the payment cannot be submitted to the site"
+        )
+
+        static let description = NSLocalizedString(
+            "The payment hasn't been posted to the store due to an error.",
+            comment: "Error description. Presented to users after collecting a payment in the app succeeds but the payment cannot be submitted to the site"
+        )
```

Comment:
I like your revised wording here, Josh. 
Two things I’m not clear about, though.
* Will people see this before the “payment successful” dialog, like in the test scenario? Or will they not see the success message at all?
* I don’t know what “The payment was authorised” means. I can take a vague guess but don’t know enough about the chain of events during payment capture. I doubt many merchants would know this, either. In a negative state of mind, my reaction is, “Well, if the payment was authorised, why couldn’t you just process it, then?”. To reduce confusion, I wonder if it might be better to leave that bit out?

Reply:
> Will people see this before the “payment successful” dialog, like in the test scenario? Or will they not see the success message at all?

@joe-keenan Merchants see the "Processing payment" dialog before this retry alert, this is a recording of the current behavior (without the changes in this PR):


https://user-images.githubusercontent.com/1945542/169990190-ac6c81aa-0829-4b5a-ae25-d0fcbcee6bbb.mov

> `The payment was authorised, but the customer was not charged due to an error.`

@joshheald This only applies to non-Interac payments, we might want to show a different message for Interac payments? I'm really not an expert on copy, please feel free to suggest one! Agree with Joe on the "authorized" part, merchants might not understand that this is a multi-step process and they just want to know what the retry might achieve and what the current state entails.

Reply:
> Can we end up in this state for Interac transactions?
> 
> I didn't think we could, because the authorize and capture happen in a single step.

Yes, we still call `capture_terminal_payment` after a successful Interac payment capture on the client side. From the repro steps in pdfdoF-Ip-p2, if the `capture_terminal_payment` request fails then the payment is captured for Interac with the order being "out of sync"/"incomplete" (the order isn't aware of the Interac payment at least in the API).

Reply:
I will update the copy altogether after hearing back about Interac from @joshheald 🙂 So far, the latest copy for non-Interac payments:

> Title: Payment capture failed
> Message: The customer was not charged due to an error.

I'd suggest updating the message for Interac payments like:

> Title: Payment recording failed
> Message: The customer was charged successfully, but the payment was not recorded in the store due to an error.

Reply:
> Merchants see the "Processing payment" dialog before this retry alert, this is a recording of the current behaviour (without the changes in this PR)

I think this is working fine. I couldn’t replicate the behaviour in the app center build, so I was just checking based on the scenario you described. The one where you wrote:

> Connect to a reader and pay with an eligible card –> the reader should show the payment as successful
Remove card –> the app should show a modal about the failure as in the "after" screenshot below

ie. Connect > Request & Take Payment > Payment Success Message > Remove Card > Failure Message
Is this the actual behaviour? I’m just confused about whether we’re going to show that the payment was successful (successful authorisation, maybe?), and then only show the failure message afterwards.

Sorry about all the questions! I know I’m getting sidetracked a bit, but I joined after all this IPP was mostly built, so I’m still learning the finer details.

> If the capture step fails repeatedly, the merchant will need to explain to the customer that they've not been charged even though there was nothing wrong with the authorisation. 

Related to my previous question, if the authorisation is successful, does that mean the customer’s card was successfully charged – and the reason we want to retry is to log the charge to Stripe (or is it our system?). I think I got confused. I thought by retrying we’d be asking the customer to re-tap their card… but we’re not doing that, are we? …or are we?

> All this can be tricky!

You’re not wrong!! 🙃

Reply:
> Related to my previous question, if the authorisation is successful, does that mean the customer’s card was successfully charged

No, the authorization happens before the customer's card is charged.

Authorization is necessary, but not sufficient, to charge the card.

Successful authorization means that the bank have said to Stripe "you're fine to charge that transaction to this card". Stripe have told us this. We still need to tell Stripe to "capture" that charge on our merchant's behalf.

> and the reason we want to retry is to log the charge to Stripe (or is it our system?). 

Yes, but also to perform the capture step. This happens when we call WCPay on the server, and then WCPay call Stripe to say "capture this charge".

> I think I got confused.

Don't worry, and do keep asking until I manage to explain it in a way that makes sense to you!

> I thought by retrying we’d be asking the customer to re-tap their card… but we’re not doing that, are we? …or are we?

No, we're retrying the capture step on the original authorization. 

The authorization was the result of the customer tapping their card, so they don't need to do it again. We're just retrying a network request.

Reply:
@jaclync OK, I think I see now...

For `card_present` payments, a failure here means three things:

1. The card was authorized
2. The details of that authorisation were not posted to WCPay
3. So the customer wasn't charged.

For `interac_present` payments we can fail here but don't end up in exactly the same state, instead:

1. The card was authorised and the customer **was charged**
2. The details of the charge weren't posted to WCPay
3. So WCPay doesn't have the order marked as paid (webhooks notwithstanding)

For `interac_present` payments, it seems important that we:

 * keep trying to pass the charge data to WCPay.
 * prevent the merchant from charging the customer again.

If we need user interaction to do the retry, I think your copy is good but maybe a slight change I'd suggest:

> Title: Payment not saved
> Message: The customer was charged, but the payment was not saved in the store due to an error.
> CTA: Retry save

Not sure whether that's really better or not though.

If we could avoid even surfacing this error and just retry in the background a few times with an increasing backoff delay, that might be even better?

In order to prevent the merchant from double charging, are we still planning to mark the order as completed locally as in the discussion pdfdoF-Ip-p2?

Reply:
To add to the complexity, also consider the case where the `capture` call makes it to the store, but the connection gets lost before receiving a response. In that case, everything is fine on the backend but the app has outdated info and can't really know if the request went through.

I like the idea of automatically retrying, because the messaging would be ambiguous. I'd also suggest we test what happens in the case I mention: if we retry a capture request that already succeeded, will the app show an error? Or will it silently realize that everything is fine? 

While looking at stripe docs the other day, I found out they have a system for [idempotent requests](https://stripe.com/docs/api/idempotent_requests), which is something we might draw inspiration from:

> Stripe's idempotency works by saving the resulting status code and body of the first request made for any given idempotency key, regardless of whether it succeeded or failed. Subsequent requests with the same key return the same result, including 500 errors.

Reply:
> In order to prevent the merchant from double charging, are we still planning to mark the order as completed locally as in the discussion pdfdoF-Ip-p2?

@joshheald Yes, this was done in https://github.com/woocommerce/woocommerce-ios/pull/6818 so the chance of double-charging in this Interac scenario should be reduced.

@joshheald @joe-keenan @koke from all the comments above, it sounds like the retry UI could be too confusing for anyone who doesn't know the inner workings of IPP including the merchants. I agree that automatic retry might be the best way to handle errors at this step, possibly using one of [Alamofire's retry policy](https://alamofire.github.io/Alamofire/Classes/RetryPolicy.html).

In terms of the next steps for this PR, do you think we close this PR and create a separate iOS issue for automatic retry (and testing the scenario that Koke mentioned)? Is there anything else we should work on for https://github.com/woocommerce/woocommerce-ios/issues/6338 for the Canada launch?

Reply:
I'm going ahead and closing this PR with an updated subtask for https://github.com/woocommerce/woocommerce-ios/issues/6338 🙂

Reply:
> Don't worry, and do keep asking until I manage to explain it in a way that makes sense to you!

It makes a lot more sense now, thanks for your patient explanation!

As for the outcome of this discussion, I think you’ve come to a good conclusion with automatic retrying.